### ISSUE TITLE
Implement "A Generic Coordinate Descent Framework for Learning from Implicit Feedback"

### ISSUE THREAD

**Issue by saulvargas** [RankSys/RankSys#38]
https://arxiv.org/abs/1611.04666v1

**Comment by fajieyuan** [comment]
I read this paper for almost three weeks and I think I can fully understand it but I cannot reproduce.
Do you want to discuss since I have my re-implementation but the performance is  far away from BPR

**Comment by fajieyuan** [comment]
I attached the code here, hope you can help me if you can find an error (can be run on librec):
// Copyright (C) 2014 Guibing Guo
//
// This file is part of LibRec.
//
// LibRec is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// LibRec is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with LibRec. If not, see <http://www.gnu.org/licenses/>.
//

package com.gla.uni.cd;

import java.util.List;

import com.gla.uni.sirgir15.TestGradient.GradPerc;
import com.google.common.base.Preconditions;

import librec.data.DenseMatrix;
import librec.data.DenseVector;
import librec.data.SparseMatrix;
import librec.data.SparseVector;
import librec.fajie.MProUtil;
import librec.intf.IterativeRecommender;
import librec.util.Randoms;
import librec.util.Strings;

/**
 * 
 * Immanuel Bayer, <strong>A Generic Coordinate Descent Framework for Learning from Implicit Feedback</strong>.
 * 
 * @author 
 * 
 */
public class ICD extends IterativeRecommender {

	private int lossf;
	private double alpha_p=1,alpha_n=0.1 ;
	
	public ICD(SparseMatrix trainMatrix, SparseMatrix testMatrix, int fold) {
		super(trainMatrix, testMatrix, fold);
		isRankingPred = true;
	}
	
	@Override
	protected void initModel() throws Exception {
		super.initModel();//Gaussian distribution (0, 0.1)
		userCache = trainMatrix.rowCache(cacheSpec);
		itemCache = trainMatrix.columnCache(cacheSpec);
	}

	@Override
	protected void buildModel() throws Exception {
		
		for (int iter = 1; iter <= numIters; iter++) {
			loss = 0;
		
			for (int f = 0; f < numFactors; f++) { //f* in the paper
				
				//w_cf=P.get(u, f), h_if=Q.get(i, f)
				//update w_cf
				double[] J_iff_arr=new double[numFactors];// JI(f,f')
				double[] J_cff_arr=new double[numFactors];// JC(f,f')
				//calculate JI(f,f')
				for (int f_ = 0; f_ < numFactors; f_++) {//f in the paper
					double J_iff_=0;
					for (int i = 0; i < numItems; i++) {
						J_iff_+= Q.get(i, f)*Q.get(i, f_);// eq(12)
					}
					J_iff_arr[f_]=J_iff_;
				}
				//here the context is the user, i.e. u=c
				for (int u = 0; u < numUsers; u++) {//for c* \in C in Algorithm 2 Line 8
					
					SparseVector pu = userCache.get(u);// positive items for u 
					int[] is = pu.getIndex();// positive item array
					
					double L_OneDerivative=0, L_TwoDerivative=0;
					double R_OneDerivative=0, R_TwoDerivative=0;
					double vals=0;
					
					for (int pos = 0; pos < is.length; pos++) {
						double x_upos=predict(u, pos);
						L_OneDerivative+=2*(alpha_p-alpha_n)*(x_upos-alpha_p/(alpha_p-alpha_n))*Q.get(pos, f);//eq.(2)
						L_TwoDerivative+=2*(alpha_p-alpha_n)*Q.get(pos, f)*Q.get(pos, f);//eq.(3)
						//caculate loss, not important for model training
						vals=(alpha_p-alpha_n)*(x_upos-alpha_p/(alpha_p-alpha_n))*(x_upos-alpha_p/(alpha_p-alpha_n));
						loss+=vals;
					}
					L_OneDerivative+=2*regU * P.get(u, f);//regularization, empirically can be omited
					L_TwoDerivative	+=2*regU;
					for (int f_ = 0; f_ < numFactors; f_++) {
						R_OneDerivative+=2*J_iff_arr[f_]*P.get(u, f_);//eq.(18)
					}
					R_TwoDerivative=2*J_iff_arr[f];//eq.(19)
					//update
					double grad=(L_OneDerivative+alpha_n*R_OneDerivative)/(L_TwoDerivative+alpha_n*R_TwoDerivative);
					P.add(u, f, -lRate * grad);//algorithm2 Line11
				}
				//calculate R(\Theta) loss, does not affact the training result
				for (int f_ = 0; f_ < numFactors; f_++) {
					double vals=0;
					for (int u = 0; u < numUsers; u++) {
						vals += P.get(u, f)* P.get(u, f_);//eq(12)
					}
					loss+=vals*J_iff_arr[f_];
				}
				
				
				//Repeat i
				for (int f_ = 0; f_ < numFactors; f_++) {
					double J_cff_=0;
					for (int u = 0; u < numUsers; u++) {
						J_cff_+= P.get(u, f)*P.get(u, f_);
					}
					J_cff_arr[f_]=J_cff_;
				}
				for (int i = 0; i < numItems;  i++) {
					SparseVector qi = itemCache.get(i);
					int[] us = qi.getIndex();// users who have rated item i
					
					double L_OneDerivative=0, L_TwoDerivative=0;
					double R_OneDerivative=0, R_TwoDerivative=0;
					double vals=0;
					for (int u = 0; u < us.length; u++) {
						double xui=predict(u, i);
						L_OneDerivative+=2*(alpha_p-alpha_n)*(xui-alpha_p/(alpha_p-alpha_n))*P.get(u, f);
						L_TwoDerivative+=2*(alpha_p-alpha_n)*P.get(u, f)*P.get(u, f);
						//caculate loss
						vals=(alpha_p-alpha_n)*(xui-alpha_p/(alpha_p-alpha_n))*(xui-alpha_p/(alpha_p-alpha_n));
						loss+=vals;
					}
					L_OneDerivative+=2*regI * Q.get(i, f);//regularization, empirically can be omited
					L_TwoDerivative+=2*regI;
					for (int f_ = 0; f_ < numFactors; f_++) {
						R_OneDerivative+=2*J_cff_arr[f_]*Q.get(i, f_);
					}
					R_TwoDerivative=2*J_cff_arr[f];
					double grad=(L_OneDerivative+alpha_n*R_OneDerivative)/(L_TwoDerivative+alpha_n*R_TwoDerivative);
					Q.add(i, f, -lRate * grad);
				}
				
				//calculate R(\Theta) loss, does not affact the training result
				for (int f_ = 0; f_ < numFactors; f_++) {
					double vals=0;
					for (int i = 0; i < numItems;  i++) {
						vals += Q.get(i, f)* Q.get(i, f_);//eq(12)
					}
					loss+=vals*J_cff_arr[f_];
				}
			
			}

			if (isConverged(iter))
				break;
		}
	}

	@Override
	public String toString() {
		return Strings.toString(new Object[] { lossf,binThold, numFactors, initLRate,
				maxLRate, regU, regI, numIters }, ",");
	}

}

**Comment by saulvargas** [comment]
Hi @fajieyuan 

Thanks for dropping by!

Indeed, I read the paper a couple of weeks as well. It looks quite elegant and promising. I have not implemented in yet, but I would like to have a performing implementation in RankSys for factorisation machines, although a good old MF version would be very welcomed.

I would not mind taking a look at your code, and maybe adapt it to work with RankSys, where it should be much easier for me to check its performance and correctness. What do you think?

Cheers
Saúl

**Comment by fajieyuan** [comment]
Hi Saúl，

Thanks for your reply.
Indeed, it looks a very interesting paper.
I totally agree your suggestion. I think before re-implementing FM+CD, a MF version is a good trial for testing the performance. If we succeed in MF version.
I think FM would be close at hand.
You can check my code, which is very clear, short and easy to understand if you follow algorithm2 (I have added notes). But I cannot find the problem in my code as the performance is not ideal according to the paper. I emailed the first author and the last author of this paper two weeks before. Immanuel gave me some general suggestions while Steffen  Rendle  has not replied to me (I believe Steffen should be very busy).

It is great you would like to take a look at the code and maybe you can easily find the bugs. The code can be run by librec http://www.librec.net/
If both of us think the code is right, then maybe we should email the author again to get some help from the original author.

Thanks
Fajie

**Comment by saulvargas** [comment]
Hi @fajieyuan 

Finallly I have some time today to give it a go to this algorithm. Were you able to make any progress with it? I will try to have a minimally working version for RankSys today.

**Comment by fajieyuan** [comment]
Hi Saúl,
thanks for your reply,
I am currently working on a paper to catch the deadline. Very sorry, I probably do not have time to work on this method at this time. Maybe later.

Best,
Fajie

**Comment by saulvargas** [comment]
No worries, I'll let you know if I make any progress.

Good luck with your deadline!

**Comment by saulvargas** [comment]
So... done!

The code is in the "icd" branch, more specifically:
https://github.com/RankSys/RankSys/blob/icd/RankSys-recommenders/src/main/java/org/ranksys/recommenders/mf/icd/ICDFactorizer.java

Will I be able of implement the version for factorisation machines today? Let's see

**Comment by hangelwen** [comment]
Hi @saulvargas , do you have benchmark of your iCD implementation with other methods regarding to recommendation performance and running time? Thanks

**Comment by saulvargas** [comment]
Hi @hangelwen 

Sorry for the massive delay in answering, my mail filter hid my github notifications!

Just in case it is still of any use to you or anyone, here is a very quick comparison of the iCD method against others that I have implemented in RankSys with the MovieLens1M data:

alg | time | prec@10
--|--|--
hkv | 1m20s | 0.29
lda | 35s | 0.25
fm-rmse | 2m50s | 0.25
pzt | 20s | 0.29
icd | 1m | 0.28
plsa | 30s | 0.26
fm-bpr | 1m40s | 0.25

Please note that I did not optimise the parameters in most cases, that fm-rmse and fm-bpr are single threaded (rest are parallel), and that the performance could drastically change depending on the dataset and your evaluation protocol
